src/primitives.rs

```diff
@@ -0,0 +1,161 @@
+use super::requests;
+
+use crate::requests::RequestApiVersions;
+use bytes::Buf;
+use core::borrow::{Borrow, BorrowMut};
+use std::fmt::{Error, Formatter};
+use DecodeErrors::BadData;
+
+enum ApiKeys {
+    Metadata = 3,
+    ApiVersions = 18,
+}
+
+pub enum ApiVersions {
+    Version0,
+    Version1,
+    Version2,
+}
+
+#[derive(Debug)]
+pub enum DecodeErrors {
+    BufferUnderflow,
+    BufferOverflow,
+    BadData,
+}
+
+pub trait Message {
+    fn get_size(&self) -> i32;
+}
+
+pub trait FromByte: Default {
+    fn decode(&mut self, buf: &mut Buf) -> Result<(), DecodeErrors>;
```

Comment:
I'm not sure if I understand the utility of having a `decode` that mutates `self` --- that seems weird to me. I would have just had this return `Result<Self, DecodeError>` and gotten rid of the separate `decode_new` function. In that case, we also wouldn't need to have a `Default` bound.

Reply:
It is the original design of `kafka-rust` : https://github.com/spicavigo/kafka-rust/blob/c58cf5f30b35fad6ab163416d51d2b99a30da9c2/src/codecs.rs#L132

Making `self` mutable is confusing, however, this can make our codes briefer:
```
request.header.decode(buf);
```
If the method returns `Result<Self, Err>`, then the code will be like:
```
request.header = request.header.decode(buf)
```

Or maybe we can return `Result<Self, Err>` and **remove** the temp variables as you suggested below. I'll try and I think it is more "rusty".

One more question is about the return type: if it returns `Result<Self, Err>` then maybe it may COPY the entire struct, and I'm wondering about passing a **reference** rather than value to improve performance. However, I don't know how to control the lifetimes of the result unless passing a `mutable &self` as a parameter. Does it matter if it returns a reference or not?

Reply:
I don't think the function to decode a Kafka type from bytes should take a self parameter. I'd expect it to look like more like this:
```rust
pub trait FromBytes {
    fn decode(buf: &mut Buf) -> Result<(), DecodeError>;
}
```
You would then use it like this:
```rust
request.header = RequestHeader::decode(buf);
```

(`Type::function` is how you call a function on a type, rather than a method on an _instance_ of that type. This is the Rust equivalent of a `static` class method in Java)

Reply:
Then we need to consider the type of the variable when decoding, say

```
header.size = i32::decode(buf);
header.api_key = i16::decode(buf);
```
instead of 
```
header.size.decode(buf);
header.api_key.decode(buf);
```

Any idea on how to detect and decode according to the type **dynamically**?

Reply:
How about we add a new function:

```
pub fn decode<T: FromByte>(&mut buf: Buf, &mut dst: T) {
    dst = T::decode(buf);
}
--------------
// Then we can do the following without specifying the type of the dst:
decode(buf, header.size);
decode(buf, header.api_key);
```

Reply:
I read a related example on the Rust tutorial book at https://doc.rust-lang.org/book/ch02-00-guessing-game-tutorial.html :
```
    let guess: u32 = guess.trim().parse()
        .expect("Please type a number!");
```

So `parse()` knows the type to be parsed, right? Then maybe we can do

```
header.size = buf.decode();
header.api_key = buf.decode();
```
where `buf` implements the `FromByte` trait. We might need to change `FromByte` a little bit to make it happen.

Reply:
This is more or less what I would recommend --- looking at `FromStr`/`parse` in the standard library is probably a good call.

 I think it's more idiomatic to build the header _from_ calls to `decode`, though, rather than constructing it with invalid values and mutating them, though. So I'd write something like this:
```rust
Header {
    size: buf.decode()?,
    api_key: buf.decode()?,
   ...
}
```
Type inference will figure out the type being decoded and dispatch to the correct implementation of `FromByte` for you.